### PR TITLE
Fix gcs_server get invalid local address

### LINKED ISSUE #7837: GCS Resolves to Wrong Address
### What is the problem?
GCS Resolves Hostname to incorrect address on MacOS. This causes a variety of problems with actors and ray functionality.

*Ray version and other system information (Python version, TensorFlow version, OS):*
Ray: 0.84
OS: MacOS Catalina

### Reproduction (REQUIRED)
Please provide a script that can be run to reproduce the issue. The script should have **no external library dependencies** (i.e., use fake or mock data / environments):
```
import ray
ray.init()
@ray.remote
class Actor:
    def G(self):
        return 30
act = Actor.remote()
ray.get(act.G.remote())
```
If we cannot run your script, we cannot fix your issue.

- [x] I have verified my script runs in a clean environment and reproduces the issue.
- [x] I have verified the issue also occurs with the [latest wheels](https://ray.readthedocs.io/en/latest/installation.html).

### PR DESCRIPTION
**Pull request by ffbin**
## Why are these changes needed?
This PR solves two problems:
1.If resolve local ip failed due to permission issues, which takes 5 seconds, so we connect google public dns server and get local ip.
2.Resolve local ip maybe get valid ip, so we add a check to make sure the address is valid.

## Related issue number

Closes #7837

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

**Comment by raulchen** [comment]
what do the invalid address look like? It seems weird to use ping to validate the address

File: src/ray/gcs/gcs_server/gcs_server.cc [hunk 1]
```diff
@@ -181,6 +185,11 @@ void GcsServer::StoreGcsServerAddressInRedis() {
   RAY_LOG(INFO) << "Finished setting gcs server address: " << address;
 }
 
+bool GcsServer::Ping(const std::string &ip, int port) {
+  AsyncClient client;
+  return client.Connect(ip, port, 1000);
```
Comment (ijrsvt) [inline src/ray/gcs/gcs_server/gcs_server.cc:190]: Could you replace this hardcoded `1000 ms` with `RayConfig::instance().internal_gcs_service_connect_wait_milliseconds()`

**Comment by ijrsvt** [comment]
@raulchen there seems to be some cases on MacOS where the given hostname leads to either failed DNS resolution or incorrect resolution.

**Comment by ijrsvt** [comment]
@ffbin Is there any way to make the logic of the AsyncClient a bit more simple--as in this [example](https://www.boost.org/doc/libs/1_53_0/doc/html/boost_asio/example/timeouts/blocking_tcp_client.cpp)?

File: src/ray/gcs/gcs_server/gcs_server.cc (hunk 1 above)
Comment (ffbin) [reply src/ray/gcs/gcs_server/gcs_server.cc]: fixed, thanks

**Comment by ffbin** [comment]
@ijrsvt Hi, I implemented it by referring to this example and simplified it, thanks.

**Comment by edoakes** [comment]
This seems to fix my local issue, but now `ray.init()` is consistently taking a long time, something like 5-10s. Can you summarize the problem we're trying to solve here? This pinging solution seems like overkill.

**Comment by ffbin** [comment]
@edoakes Hi, i am confused that a ping gcs server timeout is 100ms and will break if the ip is valid. The timeout for GCS client to connect GCS service is 5S. Can you help me see where it takes time? Thanks

File: src/ray/gcs/gcs_server/gcs_server.cc [hunk 2]
```diff
@@ -143,36 +147,52 @@ std::unique_ptr<rpc::ObjectInfoHandler> GcsServer::InitObjectInfoHandler() {
...
-  boost::asio::ip::tcp::resolver::query query(
-      boost::asio::ip::host_name(), "",
-      boost::asio::ip::resolver_query_base::flags::v4_mapped);
+  boost::asio::io_service io_service;
+  boost::asio::ip::tcp::socket socket(io_service);
   boost::system::error_code error_code;
-  boost::asio::ip::tcp::resolver::iterator iter = resolver.resolve(query, error_code);
-  boost::asio::ip::tcp::resolver::iterator end;  // End marker.
+  socket.connect(boost::asio::ip::tcp::endpoint(
+                     boost::asio::ip::address::from_string("8.8.8.8"), 53),
```
Comment (edoakes) [inline src/ray/gcs/gcs_server/gcs_server.cc:154]: Can we put `8.8.8.8` in `ray/common/constants.h` now that it's used both here and in the raylet?

**Review by edoakes — APPROVED** [review]
LGTM. Tested locally and it fixes the resolution issue.

File: src/ray/gcs/gcs_server/gcs_server.cc (hunk 2 above)
Comment (ffbin) [reply src/ray/gcs/gcs_server/gcs_server.cc]: fixed, thanks

File: src/ray/gcs/gcs_server/gcs_server.cc [hunk 3]
```diff
@@ -143,36 +147,53 @@ std::unique_ptr<rpc::ObjectInfoHandler> GcsServer::InitObjectInfoHandler() {
...
+        RAY_LOG(WARNING) << "Failed to resolve ip address, error = "
+                         << strerror(error_code.value());
+        iter = end;
+      }
+
+      if (iter != end) {
+        address =
+            primary_endpoint.address().to_string() + ":" + std::to_string(GetPort());
+      }
+    }
```
Comment (raulchen) [inline src/ray/gcs/gcs_server/gcs_server.cc:196]: I think this is too verbose to put it here. Can we wrap it in a util function?

Comment (ffbin) [reply src/ray/gcs/gcs_server/gcs_server.cc]: fixed

**Review by raulchen — APPROVED** [review]
